### PR TITLE
add interface-stream-muxer, benchmarks, and improve performance

### PR DESCRIPTION
**Pull request by jacobheun**
## Things in this PR
- [X] Be `interface-stream-muxer` compliant
- [x] Add benchmarks.
  - [x] Bonus for benchmarks against `libp2p-mplex`
- [x] Implement review feedback from https://github.com/libp2p/pull-mplex/pull/4#pullrequestreview-200815189
  - [x] Clean up closure logic for better V8 optimizations
  - [x] Change `reset` to `destroy` and verify callback needs
  - [x] `Channel.sendMsg`, verify callback needs
  - [x] Use `Buffer.allocUnsafe` when possible
  - [x] Dont use `Buffer.copy`

### CODE REVIEW

**Comment by jacobheun** [comment]
@mcollina I replicated the benchmarks from tentacoli here, doing a basic ping pong connection. I set it up to be able to check `pull-mplex` and `libp2p-mplex`. I'm going to work on the performance updates now.

I was using clinic.js to analyze the data but ran into an issue attempting to use `clinic flame`, `clinic doctor` works fine. I'll keep looking at it, but if you have any insight that would be helpful.
```sh
$ clinic flame -- node bench.js --lib=pull-mplex
To generate the report press: Ctrl + C
net.js:268
    err = this._handle.open(fd);
                       ^

Error: EEXIST: file already exists, uv_pipe_open
    at new Socket (net.js:268:24)
    at Object.<anonymous> (/Users/jacobheun/.nvm/versions/node/v10.15.1/lib/node_modules/clinic/node_modules/0x/lib/preload/redir-stdout.js:4:16)
    at Module._compile (internal/modules/cjs/loader.js:689:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:700:10)
    at Module.load (internal/modules/cjs/loader.js:599:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:538:12)
    at Function.Module._load (internal/modules/cjs/loader.js:530:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at Module._preloadModules (internal/modules/cjs/loader.js:812:12)
    at preloadModules (internal/bootstrap/node.js:599:7)
1 null
Analysing data
Generated HTML file is 88117.clinic-flame.html
You can use this command to upload it:
clinic upload 88117.clinic-flame
```

File: benchmarks/bench.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+})
+
+function buildPingPong (cb) {
+  let child
+  let dialer
+  const mplex = require(argv.lib || 'pull-mplex')
+
+  if (argv.child) {
+    child = childProcess.fork(path.join(__dirname, 'mplex-echo.js'), {
+      stdio: 'inherit'
```
Comment (mcollina) [inline benchmarks/bench.js]: I think this is what is causing clinic to crash.

**Review by mcollina — COMMENTED** [review]
Another way you can void this is to start the child process separately. You can run clinic only on one Node.js process anyway.

File: benchmarks/bench.js (hunk 1 above)
Comment (jacobheun) [reply benchmarks/bench.js]: That makes sense, I can just consolidate the test to a single file, it doesn't really need to be separated.

File: src/mplex.js [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+   * @param {Message} msg
+   */
+  _handle (msg) {
+    this._log('_handle', msg)
+    const { id, type, data } = msg
+    switch (type) {
+      // Create a new stream
+      case Types.NEW: {
+        const chan = this._newStream(id, false, true, data.toString(), this._inChannels)
+        nextTick(() => this.emit('stream', chan, id))
```
Comment (mcollina) [inline src/mplex.js]: You can avoid using a closure here by using a top-level function and passing in arguments, like so: `nextTick(emitStream, that, chan, id)`.

File: src/mplex.js [hunk 3]
```diff
@@ -0,0 +1,313 @@
...
+        const list = type & 1 ? this._outChannels : this._inChannels
+        const chan = list[id]
+        if (chan) {
+          chan.reset()
+        }
+        break
+      }
+
+      default:
+        this.emit('error', new Error('Invalid message type'))
```
Comment (mcollina) [inline src/mplex.js]: Should be emitted with nextTick

File: src/muxer.js [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+  // method added to enable pure stream muxer feeling
+  newStream (callback) {
+    callback = callback || noop
+    let stream = this.multiplex.createStream()
+    // install default error handler so that it doesn't throw
+    stream.on('error', (err) => {
+      log.err('initiator stream errored', err)
+    })
+    const conn = new Connection(stream, this.conn)
+    nextTick(() => callback(null, conn))
```
Comment (mcollina) [inline src/muxer.js]: You can do `nextTick(callback, null, conn)`.

**Review by mcollina — COMMENTED** [review]
Can you do a `clinic upload`  of doctor, flame and bubble graphs?

**Comment by jacobheun** [comment]
Here are the uploads. I did end up needing to use a smaller run for the bubble graph as the default benchmark ran into memory issues with `clinic bubbleprof`.
[`clinic doctor`](https://upload.clinicjs.org/public/1d91200a5d9bcb74fb6341d613d0aa7ccf34e9fbdd44d290f97354a0c66b7c1f/44988.clinic-doctor.html)
[`clinic flame`](https://upload.clinicjs.org/public/e1554590247e688eabcc0513e15bef7933893f5fe8939862a943f5152a37bba2/45169.clinic-flame.html)
[`clinic bubbleprof`](https://upload.clinicjs.org/public/b4d80ea6dba4c6fcf1056addeeacf798d6996c4ae860e1afb2a6ef6e35813729/46687.clinic-bubbleprof.html)

**Comment by jacobheun** [comment]
@vasco-santos this is ready for a review. I've created https://github.com/libp2p/pull-mplex/issues/7 to track release progress through to js-ipfs.

File: examples/getting-started/dialer.js [hunk 5]
```diff
@@ -0,0 +1,24 @@
+'use strict'
+
+const mplex = require('pull-mplex')
+const tcp = require('net')
+const pull = require('pull-stream')
+const toPull = require('stream-to-pull-stream')
+
+// Connect to the listener server
+const connection = tcp.connect(9999, function () {
```
Comment (vasco-santos) [inline examples/getting-started/dialer.js]: What about having the shared `9999` port in a `constants.js`?

File: src/index.js [hunk 6]
```diff
@@ -1,313 +1,43 @@
...
-          chan.reset()
-        }
-        break
-      }
-
-      default:
-        this.emit('error', new Error('Invalid message type'))
-    }
-  }
+function create (conn, isListener) {
```
Comment (vasco-santos) [inline src/index.js:20]: add jsdocs here

**Review by vasco-santos — APPROVED** [review]
LGTM! Great work! 🎉 

The example and the benchmarks are great!

Just 2 small things reported

**Pull request merged**